README.md
# pdfyoapi

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

**Note:** Replace ```Mark Kasson``` ```mkasson``` ```www.pdfyo.com``` ```mark.kasson@docsmit.com``` ```pdfyo``` ```pdfyoapi``` ```Create PDFs with the PDFYo API and mail them through the USPS with Docsmit.com``` with their correct values in [README.md](README.md), [CHANGELOG.md](CHANGELOG.md), [CONTRIBUTING.md](CONTRIBUTING.md), [LICENSE.md](LICENSE.md) and [composer.json](composer.json) files, then delete this line. You can run `$ php prefill.php` in the command line to make all replacements at once. Delete the file prefill.php as well.

This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what
PSRs you support to avoid any confusion with users and contributors.

## Structure

If any of the following are applicable to your project, then the directory structure should follow industry best practises by being named the following.

```
bin/        
config/
src/
tests/
vendor/
```


## Install

Via Composer

``` bash
$ composer require pdfyo/pdfyoapi
```

## Usage

``` php
$skeleton = new Pdfyo\PdfyoAPI();
echo $skeleton->echoPhrase('Hello, League!');
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email mark.kasson@docsmit.com instead of using the issue tracker.

## Credits

- [Mark Kasson][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/pdfyo/pdfyoapi.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/pdfyo/pdfyoapi/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/pdfyo/pdfyoapi.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/pdfyo/pdfyoapi.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/pdfyo/pdfyoapi.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/pdfyo/pdfyoapi
[link-travis]: https://travis-ci.org/pdfyo/pdfyoapi
[link-scrutinizer]: https://scrutinizer-ci.com/g/pdfyo/pdfyoapi/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/pdfyo/pdfyoapi
[link-downloads]: https://packagist.org/packages/pdfyo/pdfyoapi
[link-author]: https://github.com/mkasson
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `pdfyoapi` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


PDFYo.php
<?php

namespace Pdfyo\PdfyoAPI\Http\Controllers;

/**
 * @license Docsmit API PHP SDK
 * (c) 2014-2015 Docsmit.com, Inc. http://www.docsmit.com
 * License: GNU GPL 3
 */
class DocsmitAPI {

    private $username;
    private $password;
    public $testing;
    public $html;
    public $htmlURL;
    public $fields;
    public $JSON;
    public $mail;
    public $email;
    private $headers;
    private $curl_info;
    private $response_body;
    private $http_code;

    const HOST = "https://www.pdfyo.com/api";
    const HTTP_OK = 200;
    const HTTP_UNAUTHORIZED = 401;
// SendType values
    const ST_CERTERR = "Certified, Electronic Return Receipt";
    const ST_CERTRR = "Certified, Return Receipt";
    const ST_CERT = "Certified";
    const ST_FIRST = "First Class";
    const ST_PRIORMWS = "Priority Mail with Signature";
    const ST_PRIORM = "Priority Mail";
    const ST_PRIORITY_FRE = "Priority Mail, Flat Rate Env";
// Envelope Values
    const ENV_10 = "#10";
    const ENV_FLAT = "Flat";
    const ENV_PRIORITY_FLAT = "Priority Flat";
    const ENV_PRIORITY_PADDED = "Priority Padded";
// single or double sided printing
    const SIDED_SINGLE = 1;
    const SIDED_DOUBLE = 2;

    public function __construct($username, $password, $softwareID) {
        $this->URIBase = ($URIBase != "") ? $URIBase : "https://www.pdfyo.com/api";
        $this->username = $username;
        $this->password = $password;
    }

    public static function authping($username, $password) {
        $process = curl_init(HOST . "/authping");
        curl_setopt($process, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
        curl_setopt($process, CURLOPT_HEADER, 1);
        curl_setopt($process, CURLOPT_USERPWD, $username . ":" . $password);
        curl_setopt($process, CURLOPT_TIMEOUT, 30);
//curl_setopt($process, CURLOPT_POST, 1);
//curl_setopt($process, CURLOPT_POSTFIELDS, $payloadName);
        curl_setopt($process, CURLOPT_RETURNTRANSFER, TRUE);
        $return = curl_exec($process);
        $this->http_code = $this->curl_info["http_code"];
        curl_close($process);
        return $this->http_code;
    }

    public static function health() {
        $process = curl_init(HOST . "/health");
        curl_setopt($process, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
        curl_setopt($process, CURLOPT_HEADER, 1);
//curl_setopt($process, CURLOPT_USERPWD, $username . ":" . $password);
        curl_setopt($process, CURLOPT_TIMEOUT, 30);
//curl_setopt($process, CURLOPT_POST, 1);
//curl_setopt($process, CURLOPT_POSTFIELDS, $payloadName);
        curl_setopt($process, CURLOPT_RETURNTRANSFER, TRUE);
        $return = curl_exec($process);
        $this->http_code = $this->curl_info["http_code"];
        curl_close($process);
        return $this->http_code;
    }

    public function responseBody() {
        return $this->response_body;
    }

    public function responseObject() {
        return json_decode($this->responseBody());
    }

    function curlInfo() {
        return $this->curl_info;
    }

    public function status() {
        return $this->http_code;
    }

    function returnedError() {
        return ( $this->http_code >= 400);
    }

    function returnedSuccess() {
        return ( $this->http_code == 200 || $this->http_code == 201);
    }

    public function send($params) {

        $process = curl_init(HOST . "/pdfgen");
        curl_setopt($process, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
        curl_setopt($process, CURLOPT_HEADER, 1);
        curl_setopt($process, CURLOPT_USERPWD, $username . ":" . $password);
        curl_setopt($process, CURLOPT_TIMEOUT, 30);
        curl_setopt($process, CURLOPT_POST, 1);
        curl_setopt($process, CURLOPT_POSTFIELDS, json_encode($params));
        curl_setopt($process, CURLOPT_RETURNTRANSFER, TRUE);
        $return = curl_exec($process);
        curl_close($process);
    }

    private function execAndFill($ch, $params = NULL, $contentType = "application/json", $jsonEncode = true) {
        curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type:$contentType"));
        if ($params != NULL) {
            if ($jsonEncode)
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));
            else
                curl_setopt($ch, CURLOPT_POSTFIELDS, $params);
        }
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_USERPWD, $this->token);
        $this->response_body = curl_exec($ch);

        $this->curl_info = curl_getinfo($ch);
        $header_size = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
        curl_close($ch);
        $this->headers = substr($this->response_body, 0, $header_size);
        $this->http_code = $this->curl_info["http_code"];
    }

    public function get($url, $params = NULL) {

        if ($params != NULL) {
            $qsSeparator = strrpos($url, "?") === FALSE ? '?' : '&';
            $url .= $qsSeparator . http_build_query($params);
        }

        $ch = curl_init($this->URIBase() . $url);
        $this->execAndFill($ch);
    }

    public function post($url, $params = NULL, $isFile = false) {
        $ch = curl_init($this->URIBase() . $url);
        curl_setopt($ch, CURLOPT_POST, true);

        if ($isFile && !empty($params["filePath"]) && file_exists($params["filePath"])) {
            $finfo = new finfo(FILEINFO_MIME);
            $mimeType = $finfo->file($params["filePath"]);
            $params["file"] = new CurlFile($params["filePath"], $mimeType);
            $this->execAndFill($ch, $params, "multipart/form-data", false);
        } elseif ($isFile && !empty($params["blob"])) {
            $this->execAndFill($ch, $params, "multipart/form-data", false);
        } else
            $this->execAndFill($ch, $params);

        if ($this->http_code == self::HTTP_UNAUTHORIZED) {
            $this->refreshToken();
            $ch = curl_init($this->URIBase() . $url);
            curl_setopt($ch, CURLOPT_POST, true);
            if ($isFile && !empty($params["filePath"]) && file_exists($params["filePath"])) {
                $finfo = new finfo(FILEINFO_MIME);
                $mimeType = $finfo->file($params["filePath"]);
                $params["file"] = new CurlFile($params["filePath"], $mimeType);
                $this->execAndFill($ch, $params, "multipart/form-data", false);
            } elseif ($isFile && !empty($params["blob"])) {
                $this->execAndFill($ch, $params, "multipart/form-data", false);
            } else
                $this->execAndFill($ch, $params);
        }
    }


// If you don't have PECL with http_parse_headers
//see: http://php.net/manual/en/function.http-parse-headers.php#68698
}


src/PdfyoAPIServiceProvider.php
<?php
namespace Pdfyo\PdfyoAPI;
use Illuminate\Support\ServiceProvider;
use Illuminate\Routing\Router;
class PdfyoAPIServiceProvider extends ServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        // use this if your package has views
//        $this->loadViewsFrom(realpath(__DIR__.'/resources/views'), 'pdfyoapi');
        
        // use this if your package has lang files
//        $this->loadTranslationsFrom(__DIR__.'/resources/lang', 'pdfyoapi');
        
        // use this if your package has routes
//        $this->setupRoutes($this->app->router);
        
        // use this if your package needs a config file
        // $this->publishes([
        //         __DIR__.'/config/config.php' => config_path('pdfyoapi.php'),
        // ]);
        
        // use the vendor configuration file as fallback
        // $this->mergeConfigFrom(
        //     __DIR__.'/config/config.php', 'pdfyoapi'
        // );
    }
    /**
     * Define the routes for the application.
     *
     * @param  \Illuminate\Routing\Router  $router
     * @return void
     */
    public function setupRoutes(Router $router)
    {
        $router->group(['namespace' => 'Pdfyo\PdfyoAPI\Http\Controllers'], function($router)
        {
            require __DIR__.'/Http/routes.php';
        });
    }
    /**
     * Register any package services.
     *
     * @return void
     */
    public function register()
    {
        $this->registerSkeleton();
        
        // use this if your package has a config file
        // config([
        //         'config/pdfyoapi.php',
        // ]);
    }
    private function registerSkeleton()
    {
        $this->app->bind('pdfyoapi',function($app){
            return new PDFYoAPI($app);
        });
    }
}


src/pdfyoapiServiceProvider.php
<?php
namespace Pdfyo\PdfyoAPI;
use Illuminate\Support\ServiceProvider;
use Illuminate\Routing\Router;
class PdfyoAPIServiceProvider extends ServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        // use this if your package has views
        $this->loadViewsFrom(realpath(__DIR__.'/resources/views'), 'pdfyoapi');
        
        // use this if your package has lang files
        $this->loadTranslationsFrom(__DIR__.'/resources/lang', 'pdfyoapi');
        
        // use this if your package has routes
        $this->setupRoutes($this->app->router);
        
        // use this if your package needs a config file
        // $this->publishes([
        //         __DIR__.'/config/config.php' => config_path('pdfyoapi.php'),
        // ]);
        
        // use the vendor configuration file as fallback
        // $this->mergeConfigFrom(
        //     __DIR__.'/config/config.php', 'pdfyoapi'
        // );
    }
    /**
     * Define the routes for the application.
     *
     * @param  \Illuminate\Routing\Router  $router
     * @return void
     */
    public function setupRoutes(Router $router)
    {
        $router->group(['namespace' => 'Pdfyo\PdfyoAPI\Http\Controllers'], function($router)
        {
            require __DIR__.'/Http/routes.php';
        });
    }
    /**
     * Register any package services.
     *
     * @return void
     */
    public function register()
    {
        $this->registerSkeleton();
        
        // use this if your package has a config file
        // config([
        //         'config/pdfyoapi.php',
        // ]);
    }
    private function registerSkeleton()
    {
        $this->app->bind('pdfyoapi',function($app){
            return new Skeleton($app);
        });
    }
}


src/PDFYoAPI.php
<?php

namespace Pdfyo\PdfyoAPI;

/**
 * @license Docsmit API PHP SDK
 * (c) 2014-2015 Docsmit.com, Inc. http://www.docsmit.com
 * License: GNU GPL 3
 */
class PDFYoAPI {

    private $username;
    private $password;
    public $testing;
    public $html;
    public $htmlURL;
    public $fields;
    public $JSON;
    public $mail;
    public $email;
    private $headers;
    private $curl_info;
    private $response_body;
    private $http_code;

    const HOST = "https://www.pdfyo.com/api";
    const HTTP_OK = 200;
    const HTTP_UNAUTHORIZED = 401;
// SendType values
    const ST_CERTERR = "Certified, Electronic Return Receipt";
    const ST_CERTRR = "Certified, Return Receipt";
    const ST_CERT = "Certified";
    const ST_FIRST = "First Class";
    const ST_PRIORMWS = "Priority Mail with Signature";
    const ST_PRIORM = "Priority Mail";
    const ST_PRIORITY_FRE = "Priority Mail, Flat Rate Env";
// Envelope Values
    const ENV_10 = "#10";
    const ENV_FLAT = "Flat";
    const ENV_PRIORITY_FLAT = "Priority Flat";
    const ENV_PRIORITY_PADDED = "Priority Padded";
// single or double sided printing
    const SIDED_SINGLE = 1;
    const SIDED_DOUBLE = 2;

    public function __construct($username, $password) {
        $this->username = $username;
        $this->password = $password;
    }

    public static function authping($username, $password) {
        $process = curl_init(self::HOST . "/authping");
        curl_setopt($process, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
        curl_setopt($process, CURLOPT_HEADER, 1);
        curl_setopt($process, CURLOPT_USERPWD, $username . ":" . $password);
        curl_setopt($process, CURLOPT_TIMEOUT, 30);
//curl_setopt($process, CURLOPT_POST, 1);
//curl_setopt($process, CURLOPT_POSTFIELDS, $payloadName);
        curl_setopt($process, CURLOPT_RETURNTRANSFER, TRUE);
        $return = curl_exec($process);
        $curl_info = curl_getinfo($process);
        $http_code = $curl_info["http_code"];
        curl_close($process);
        return $http_code;
    }

    public static function health() {
        $process = curl_init(self::HOST . "/health");
        curl_setopt($process, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
        curl_setopt($process, CURLOPT_HEADER, 1);
//curl_setopt($process, CURLOPT_USERPWD, $username . ":" . $password);
        curl_setopt($process, CURLOPT_TIMEOUT, 30);
//curl_setopt($process, CURLOPT_POST, 1);
//curl_setopt($process, CURLOPT_POSTFIELDS, $payloadName);
        curl_setopt($process, CURLOPT_RETURNTRANSFER, TRUE);
        $return = curl_exec($process);
        $curl_info = curl_getinfo($ch);
        $http_code = $curl_info["http_code"];
        curl_close($process);
        return $http_code;
    }

    public function responseBody() {
        return $this->response_body;
    }

    public function responseObject() {
        return json_decode($this->responseBody());
    }

    function curlInfo() {
        return $this->curl_info;
    }

    public function status() {
        return $this->http_code;
    }

    function returnedError() {
        return ( $this->http_code >= 400);
    }

    function returnedSuccess() {
        return ( $this->http_code == 200 || $this->http_code == 201);
    }

    public function send($params = null) {
        if ($params == null) {
            $params['html'] = $this->html;
            $params['testing'] = $this->testing;
            $params['mail'] = $this->mail;
            $params['fields'] = $this->fields;
            $params['email'] = $this->email;
        }

        $process = curl_init(self::HOST . "/pdfgen");
        curl_setopt($process, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
        curl_setopt($process, CURLOPT_HEADER, 1);
        curl_setopt($process, CURLOPT_USERPWD, $this->username . ":" . $this->password);
        curl_setopt($process, CURLOPT_TIMEOUT, 30);
        curl_setopt($process, CURLOPT_POST, 1);
        curl_setopt($process, CURLOPT_POSTFIELDS, json_encode($params));
        curl_setopt($process, CURLOPT_RETURNTRANSFER, TRUE);
        $this->response_body = curl_exec($process);
        $this->curl_info = curl_getinfo($process);
        $this->http_code = $this->curl_info["http_code"];
        curl_close($process);
    }

    private function execAndFill($ch, $params = NULL, $contentType = "application/json", $jsonEncode = true) {
        curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type:$contentType"));
        if ($params != NULL) {
            if ($jsonEncode)
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));
            else
                curl_setopt($ch, CURLOPT_POSTFIELDS, $params);
        }
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_USERPWD, $this->token);
        $this->response_body = curl_exec($ch);

        $this->curl_info = curl_getinfo($ch);
        $header_size = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
        curl_close($ch);
        $this->headers = substr($this->response_body, 0, $header_size);
        $this->http_code = $this->curl_info["http_code"];
    }

    public function get($url, $params = NULL) {

        if ($params != NULL) {
            $qsSeparator = strrpos($url, "?") === FALSE ? '?' : '&';
            $url .= $qsSeparator . http_build_query($params);
        }

        $ch = curl_init($this->URIBase() . $url);
        $this->execAndFill($ch);
    }

    public function post($url, $params = NULL, $isFile = false) {
        $ch = curl_init($this->URIBase() . $url);
        curl_setopt($ch, CURLOPT_POST, true);

        if ($isFile && !empty($params["filePath"]) && file_exists($params["filePath"])) {
            $finfo = new finfo(FILEINFO_MIME);
            $mimeType = $finfo->file($params["filePath"]);
            $params["file"] = new CurlFile($params["filePath"], $mimeType);
            $this->execAndFill($ch, $params, "multipart/form-data", false);
        } elseif ($isFile && !empty($params["blob"])) {
            $this->execAndFill($ch, $params, "multipart/form-data", false);
        } else
            $this->execAndFill($ch, $params);

        if ($this->http_code == self::HTTP_UNAUTHORIZED) {
            $this->refreshToken();
            $ch = curl_init($this->URIBase() . $url);
            curl_setopt($ch, CURLOPT_POST, true);
            if ($isFile && !empty($params["filePath"]) && file_exists($params["filePath"])) {
                $finfo = new finfo(FILEINFO_MIME);
                $mimeType = $finfo->file($params["filePath"]);
                $params["file"] = new CurlFile($params["filePath"], $mimeType);
                $this->execAndFill($ch, $params, "multipart/form-data", false);
            } elseif ($isFile && !empty($params["blob"])) {
                $this->execAndFill($ch, $params, "multipart/form-data", false);
            } else
                $this->execAndFill($ch, $params);
        }
    }


// If you don't have PECL with http_parse_headers
//see: http://php.net/manual/en/function.http-parse-headers.php#68698
}
